### ISSUE TITLE
tsserver.js using too much memory

### ISSUE THREAD

**Issue by Meisolsson** [microsoft/TypeScript#30034]
Issue Type: <b>Performance Issue</b>

After running the editor for a while tsserver uses more and more memory with only 3-4 files open at the time

VS Code version: Code 1.30.2 (61122f88f0bf01e2ac16bdb9e1bc4571755f5bd8, 2019-01-07T22:48:31.260Z)
OS version: Darwin x64 18.2.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-4278U CPU @ 2.60GHz (4 x 2600)|
|GPU Status|2d_canvas: enabled<br>checker_imaging: disabled_off<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: enabled<br>rasterization: enabled<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|3, 3, 3|
|Memory (System)|8.00GB (0.04GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|

</details><details>
<summary>Process Info</summary>

```
CPU %	Mem MB	   PID	Process
   15	    82	 48248	code main
    0	   106	 48249	   gpu-process
    0	    25	 48457	   shared-process
    0	   311	 58478	   window (types.js — keyruler (Workspace))
    0	     0	 58479	     /bin/bash -l
    0	    82	 58518	     extensionHost
    0	  1720	 58687	       electron_node tsserver.js 
    0	    25	 58722	         electron_node typingsInstaller.js typesMap.js 
    0	    16	 58684	     watcherService
    0	    33	 61982	     /Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper.app/Contents/MacOS/Code Helper /Applications/Visual Studio Code.app/Contents/Resources/app/out/bootstrap-fork ms-vscode.node2 {"common.vscodemachineid":"aa09937a8f20affad38b2536d7c6d608be681a06011c1b9cc89ab3faa2635ecc","common.vscodesessionid":"8bc62413-df46-45fb-a6b9-60ee4e7fd3691549297291928"} AIF-d9b70cd4-b9f9-4d70-929b-a071c400b217
    2	    74	 62679	   window (Issue Reporter)
    0	     0	 62682	   /bin/ps -ax -o pid=,ppid=,pcpu=,pmem=,command=
```

</details>
<details>
<summary>Workspace Info</summary>

```
|  Window (types.js — keyruler (Workspace))
|    Folder (keyruler): 6 files
|      File types: json(2) md(1) gitignore(1) js(1)
|      Conf files: package.json(1)
|    Folder (kafka-avro): 26 files
|      File types: js(12) yml(4) json(4) editorconfig(1) gitignore(1) nvmrc(1)
|                  md(1) opts(1)
|      Conf files: package.json(1)
|    Folder (avsc): 147 files
|      File types: js(39) json(31) avsc(20) py(6) avdl(5) md(5) png(5) java(4)
|                  proto(4) avro(3)
|      Conf files: package.json(3) launch.json(1)
|      Launch Configs: node;
```

</details>
Extensions: none

**Comment by mjbvz** [comment]
Thank you for the example project @Meisolsson. I just tested this in the [latest VS Code insiders build](https://code.visualstudio.com/insiders/) and have not been able to make memory increase past around 300mb after opening most of the files in the project and making a few edits in them

A few other questions to help us investigate this:

- Do you see this issue in the latest VS Code builds?
- DO you see the memory usage spike right away when you open the files, or only after working in the project for some time?
- Does anything in particular seem to trigger the high memory usage?

**Comment by mattch** [comment]
I'm having the same issue, though for me the cpu and memory spikes are sporadic.  Once the start, they usually won't go away until I restart vscode.   When I'm not seeing the problem, there doesn't seem to be any way to make it happen. I've tried opening a dozen large files without any spike in cpu or memory.

I have the latest vscode and typescript.

**Comment by sheetalkamat** [comment]
@mattch Please share the tsserver log and ti log using vscode insiders latest build and using [typescript nightly drop](https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Nightly%20Builds.md#visual-studio-code) when the issue repros.
You can upload logs by following these steps:
1. With "typescript.tsserver.log": "verbose" enabled
2. Restart vscode or vscode insiders
3. Reproduce the problem
4. You can share the logs using
    - either by sharing logs in the folder that gets opened by running `open tsserver log` from command palette 
    or
    - Run ```code --upload-logs``` on the command line (or ```code-insiders --upload-logs``` if using vscode insiders) and follow instructions and Then share the log file ID

**Comment by delroh** [comment]
Same issue happening to to me, on Windows 10. Haven't really tried to find a pattern of why/when it happens but what I usually do is leave the editor open for days, putting the computer to sleep and wake up after long or short time, and sometimes leaving `npm start` running for a react project.

Came here after seeing the tsserver process was using ~880MB; after writing this, went to take a screenshot and the process was there no more and VS Code went back to using ~287MB overall. Will  try @sheetalkamat 's suggestion to log the tsserver and see what's up.

**Comment by jadbox** [comment]
tsserver.js is using 1.2 gigs of memory on  my small-ish front-end project. I'm not sure how to create a reproducible setup for this. If needed, I could give a developer access to my project for testing.

![19-07-11-09-31](https://user-images.githubusercontent.com/595843/61068612-4bf5ed00-a3bf-11e9-8afe-6a66453a5cd6.png)

**Comment by jadbox** [comment]
Possibly related issue: https://github.com/Microsoft/TypeScript/issues/30981

**Comment by sheetalkamat** [comment]
@jadbox your project along with tsserver log as per https://github.com/microsoft/TypeScript/issues/30034#issuecomment-493161120 would be helpful to investigate. Thanks.

**Comment by vkaracic** [comment]
Same issue with me:
![Screenshot from 2019-09-02 16-51-09](https://user-images.githubusercontent.com/2808092/64122435-ecee9c00-cda1-11e9-8b98-e3c6cfe731a6.png)

**Comment by JamesJGoodwin** [comment]
Same issue here. Only 7 tabs of code opened.

![image](https://user-images.githubusercontent.com/17240507/65386222-3e56cf00-dd41-11e9-95fc-c55b35b0a5a3.png)

**Comment by mmahdigh** [comment]
Same here.
![tsserver](https://user-images.githubusercontent.com/46849131/66833750-3f0e0a00-ef69-11e9-821d-cb05ba55cd21.PNG)

**Version: Code** 1.39.1 (88f15d17dca836346e787762685a40bb5cce75a8, 2019-10-10T23:31:28.683Z)
**OS Version**: Windows_NT x64 10.0.17134
**CPUs:** Intel(R) Core(TM) i7-4702MQ CPU @ 2.20GHz (8 x 2195)
**Memory (System)**:  5.76GB (0.82GB free)
**VM**: 40%

**Comment by sheetalkamat** [comment]
Please share logs with typescript nightly for us to be able to investigate this. Screen shots though showing that you are facing the issue are not helpful for investigation at all.

You can upload logs by following these steps using [typescript nightly drop](https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Nightly%20Builds.md#visual-studio-code).:

1. Set `"typescript.tsserver.log": "verbose"`
2. Restart vscode or vscode insiders
3. Reproduce the problem
4. In vscode or vscode insiders, run the `TypeScript: Open TS Server log` command
5. This should open a large file called `tsserver.log`. Please share this file. Additionally you can also share file `ti.log` created next to it.

**Comment by nhducseuit** [comment]
[tsserver.log](https://github.com/microsoft/TypeScript/files/3907042/tsserver.log)
[ti-9572.log](https://github.com/microsoft/TypeScript/files/3907043/ti-9572.log)

**Comment by nhducseuit** [comment]
Anybody ?

**Comment by harveyconnor** [comment]
having the same issue, VSCode eventually becomes unusable after developing for extended hours.

**Comment by PabloHernanVila** [comment]
Same issue here. Is there any update?

**Comment by isostarec** [comment]
Same issue here as well.

**Comment by mheavenor** [comment]
Looks like this may be a duplicate of [this issue](https://github.com/microsoft/vscode/issues/85919) and will be solved with VS Code 1.43

**Comment by meotimdihia** [comment]
Vscode Insider 1.43 is still using 450MB+ for me.

**Comment by kmvan** [comment]
![image](https://user-images.githubusercontent.com/3839554/75685280-4abe5b00-5cd5-11ea-9054-8f3142adeb2a.png)

Same with me.

Watching file.

**Comment by mrdimaan** [comment]
Same here. ts-server uses all of my 24GB ram and my operating system does not work. On PopOS Linux

**Comment by SarkarKurdish** [comment]
Same here Manjaro Linux it's unusable

**Comment by mjbvz** [comment]
Closing as we need specific repos to investigate performance issues

If you believe TSServer is using too much memory when using TS 4.1.5+, please open a new issue. Include as much information about your project as possible (ideally just like to the source code so we can reproduce this ourselves)

**Comment by cristianjuarezz** [comment]
This is still an issue in 2023, my IDE is not Visual Studio Code, but still reproduceable by leaving a node server running for a while.
Sounds like a typical memory leak.

![image](https://github.com/microsoft/TypeScript/assets/84170516/133f22c6-7197-4651-81f2-2fbca6f5dc82)
![image](https://github.com/microsoft/TypeScript/assets/84170516/576480a7-8d31-4cbc-ae7e-d48d713e8212)

Also I noticed: The memory only gets freed once I kill the IDE or terminal in which i had the server, and not when i kill the server with Ctrl + C

**Comment by AlaricWeb** [comment]
I have the same problem. Is there any way to solve it

**Comment by IOLOII** [comment]
# 2024.5.4 same problem again!
<img width="660" alt="image" src="https://github.com/microsoft/TypeScript/assets/34856171/d1a54c41-f989-46ff-9901-5b6379014344">

Although it is only more than 500 now, it is the size after I forced the shutdown. When I was developing the project, the memory occupied very frequently, so I had to open the process manager and close it.

**Comment by jilherme** [comment]
I'm having 500MB+ mem usage too

![image](https://github.com/microsoft/TypeScript/assets/74925460/00b94aaf-b5f2-42c3-b8e6-92bd262b17c8)

**Comment by mjbvz** [comment]
Please open new issues instead of posting on old closed ones

Also depending on what's in your project, 500mb may be expected
